### ISSUE TITLE
Real numbers should be parsed and `visit_f64` instead of `visit_str`

### ISSUE THREAD

**Issue by jwilm** [dtolnay/serde-yaml#24]
I ran into a problem implementing a `deserialize_with` for a type based on `f32`. Source (important part here is the `impl Visitor for FloatVisitor`):

```rust
impl DeserializeFromF32 for Size {
    fn deserialize_from_f32<D>(deserializer: &mut D) -> ::std::result::Result<Self, D::Error>
        where D: serde::de::Deserializer
    {
        use std::marker::PhantomData;

        struct FloatVisitor<__D> {
            _marker: PhantomData<__D>,
        }

        impl<__D> ::serde::de::Visitor for FloatVisitor<__D>
            where __D: ::serde::de::Deserializer
        {
            type Value = f32;

            fn visit_f32<E>(&mut self, value: f32) -> ::std::result::Result<Self::Value, E>
                where E: ::serde::de::Error
            {
                Ok(value)
            }
        }

        deserializer
            .deserialize_f32(FloatVisitor::<D>{ _marker: PhantomData })
            .map(|v| Size::new(v))
    }
}
```

Strangely, the config was failing to parse saying it expected a `&str`. The problem is in the [yaml deserialize impl]:

```rust
    fn deserialize<V>(&mut self, mut visitor: V) -> Result<V::Value>
        where V: de::Visitor,
    {
        match *self.doc {
            Yaml::Integer(i) => visitor.visit_i64(i),
            Yaml::Real(ref s) |
            Yaml::String(ref s) => visitor.visit_str(s),
```

Floats end up being passed to `visit_str`! Can we change this to parse the float and pass it to `visit_f64`?

[yaml deserialize impl]: https://github.com/dtolnay/serde-yaml/blob/master/yaml/src/de.rs#L235-L240

**Comment by jwilm** [comment]
Thanks @dtolnay :smile:

**Comment by dtolnay** [comment]
:+1: yaml-rust just published the release that has been blocking serde_yaml::Value #23 so I am going to wrap that up and get 0.5.0 out as soon as I can.

**Comment by dtolnay** [comment]
I released [0.5.0](https://github.com/dtolnay/serde-yaml/releases/tag/0.5.0) with this change.
